working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+threads which all update the same global variable:
+
+```ruby
+count = 0
+threads = 5.times.map do |i|
+  puts "T#{i}: starting"
+  Thread.new do
+    count += 1
+    puts "T#{i}: done"
+  end
+end
+
+threads.each { |t| t.join }
+puts "Counted to #{count}"
+```
+
+```console
+$ ruby test.rb
+T0: starting
+T1: starting
+T2: starting
+T3: starting
+T4: starting
+T1: done
+T4: done
+T0: done
+T2: done
+T3: done
+Counted to 5
+```
+
+Concurrency in both languages is severely limited by a global lock which
+protects interpreter's integrity. This lock is known _Global Interpreter Lock_
+(GIL) in Python and _Global VM Lock_ (GVL) in Ruby. GIL/GVL ensures that an
+interpreter is only running on one thread at a time. Scheduling mechanisms built
+into the interpreter allow it to switch between threads giving each a chance to
+run concurrently. This means executions of Python/Ruby code on different threads
+are interleaved, but serialized. You can observe non-atomic behaviors and data
+races (the VM will not crash though), but you can't utilize multicore
+capabilities. CPython developers are actively exploring the possibility to
+remove the GIL see [PEP 703][].
+
+**Erlang** is a functional programming language for creating highly concurrent
+distributed systems which represents another extreme: no shared memory
+multithreading or low-level threading primitives at all. Design principles
+behind **Erlang** are summarized in Joe Armstrong's PhD thesis [Making reliable
+distributed systems in the presence of software
+errors][Joe Armstrong PhD Thesis]. Isolation between lightweight processes which
+form a running application is the idea at the very core of **Erlang**'s design,
+to quote section 2.4.3 of the thesis:
+
+> The notion of isolation is central to understanding COP, and to the
+> construction of fault-tolerant software. Two processes operating on the same
+> machine must be as independent as if they ran on physically separated machines
+>
+> ...
+>
+> Isolation has several consequences:
+>
+> 1. Processes have “share nothing” semantics. This is obvious since they are
+>    imagined to run on physically separated machines.
+> 2. Message passing is the only way to pass data between processes. Again since
+>    nothing is shared this is the only means possible to exchange data.
+> 3. Isolation implies that message passing is asynchronous. If process
+>    communication is synchronous then a software error in the receiver of a
+>    message could indefinitely block the sender of the message destroying the
+>    property of isolation.
+> 4. Since nothing is shared, everything necessary to perform a distributed
+>    computation must be copied. Since nothing is shared, and the only way to
+>    communicate between processes is by message passing, then we will never
+>    know if our messages arrive (remember we said that message passing is
+>    inherently unreliable.) The only way to know if a message has been
```

Comment:
And even that runs into the Byzantine general problem. The one who sent the conformation message doesn't know whether their confirmation message arrived. (Which is problematic with handing over ownership of resources that need to be freed.)

Reply:
We are going to deep here :)